Wrappers/Python/ccpi/web_viewer/trame_viewer.py

```diff
@@ -0,0 +1,772 @@
+
+    def change_slice_window(self, new_window):
+        self.cil_viewer.setSliceColourWindow(window=new_window)
+
+    def change_slice_level(self, new_level):
+        self.cil_viewer.setSliceColourLevel(level=new_level)
+
+    def change_window_level_detail_sliders(self, show_detailed):
+        if show_detailed == self.slice_window_sliders_are_detailed:
+            return
+        # Translate current colour level and colour window to range_min and range_max
+        current_level = self.cil_viewer.getSliceColourLevel()
+        current_window = self.cil_viewer.getSliceColourWindow()
+        range_min = (current_window - 2 * current_level) / -2  # The reverse of the window calculation in web_app.change_slice_window_level
+        range_max = current_window + range_min
+
+        # Setup the defaults pre-flip
+        self.slice_window_range_defaults = [range_min, range_max]
+        self.slice_level_default = current_level
+        self.slice_window_default = current_window
+
+        # Toggle the detailed sliders
+        self.slice_window_sliders_are_detailed = show_detailed
+
+        # Reconstruct the detailed sliders
+        self.slice_window_range_slider = self.construct_slice_window_range_slider()
+        self.slice_window_slider = self.construct_slice_window_slider()
+        self.slice_level_slider = self.construct_slice_level_slider()
+
+        # Reconstruct the drawer and push it
+        self.construct_drawer_layout()
+        update_layout(self.layout)
+
+    def change_slice_visibility(self, visbility):
+        if visbility:
+            self.cil_viewer.imageSlice.VisibilityOn()
+            self.disable_2d = False
+        else:
+            self.cil_viewer.imageSlice.VisibilityOff()
+            self.disable_2d = True
+        self.create_drawer_ui_elements()
+        update_layout(self.layout)
+        self.cil_viewer.updatePipeline()
+
+    def change_volume_visibility(self, visibility):
+        if not self.cil_viewer.volume_render_initialised:
+            self.cil_viewer.installVolumeRenderActorPipeline()
+
+        if visibility:
+            self.cil_viewer.volume.VisibilityOn()
+            self.cil_viewer.light.SwitchOn()
+            self.disable_3d = False
+        else:
+            self.cil_viewer.volume.VisibilityOff()
+            self.cil_viewer.light.SwitchOff()
+            self.disable_3d = True
+        self.create_drawer_ui_elements()
+        update_layout(self.layout)
+        self.cil_viewer.updatePipeline()
+
+    def change_background_colour(self, colour):
+        if colour == "cil_viewer_blue":
+            color_data = (.1, .2, .4)
+        else:
+            color_data = getattr(colors, colour.lower())
+        self.cil_viewer.ren.SetBackground(color_data)
+        self.cil_viewer.updatePipeline()
+
+    def change_clipping(self, clipping_on):
+        app = vuetify.get_app_instance()
+        app.set(key="slice_visibility", value=False)
+        if hasattr(self.cil_viewer, "planew"):
```

Comment:
could lines 753 to 759 be replaced by calling this?
https://github.com/Pasarus/CILViewer/blob/90914626c9e7c45e4bc057dd13c92e0eb7cb3ae9/Wrappers/Python/ccpi/viewer/CILViewer.py#L183
or is it because of the slice toggle that we can't do that?

Reply:
it would be better to call code in the viewer if possible so that if we had to change any functionality we would only have to change in one place

Reply:
I have changed much of the slice visibility parts of the work to fix the bugs around it.